File: xor-key/Sample.java
import java.io.*;
import java.util.*;

public class Sample {

	void solve(BufferedReader rd, int Q, int[] nums) throws Exception {
		for(; Q > 0; Q--) {
			StringTokenizer st = new StringTokenizer(rd.readLine());
			int a = Integer.parseInt(st.nextToken());
			int p = Integer.parseInt(st.nextToken());
			int q = Integer.parseInt(st.nextToken());
			
			int max = Integer.MIN_VALUE;
			for(int i = p - 1; i < q; i++){ 
				int xor = nums[i] ^ a;
				if(xor > max) max = xor;
				// System.out.printf("%4d ^ %4d = %4d : %10s ^ %10s = %10s", nums[i], a, xor, Integer.toBinaryString(nums[i]), Integer.toBinaryString(a), Integer.toBinaryString(xor));
				// System.out.println();
			}
			
			System.out.println(max);
		}
	}

	void run(BufferedReader rd) throws Exception {
		int T = Integer.parseInt(rd.readLine());
		for(; T > 0; T--) {
			StringTokenizer st = new StringTokenizer(rd.readLine());
			int N = Integer.parseInt(st.nextToken());
			int Q = Integer.parseInt(st.nextToken());
			int[] nums = new int[N];
			
			st = new StringTokenizer(rd.readLine());
			for(int i = 0; i < N; i++){
				nums[i] = Integer.parseInt(st.nextToken());
			}
			
			solve(rd, Q, nums);
		}
	}

	public static void main(String[] args) throws Exception {
		Reader rd = (args.length == 0) ? new InputStreamReader(System.in) : new FileReader(args[0]);
		new Sample().run(new BufferedReader(rd));
	}
}

File: hackerrank/alien-languages/Sample.scala
import java.io.{FileReader, Reader, BufferedReader, InputStreamReader}
import java.util.StringTokenizer
import scala.collection.mutable
import scala.math.max
import scala.util.control.Breaks._

object Sample {
  def solve(_rd: Reader) {
    val mod = 100000007
    val rd = new BufferedReader(_rd)
    val T = rd.readLine().toInt

    for (i <- 0 until T) {
      val st = new StringTokenizer(rd.readLine())
      val N = st.nextToken().toInt
      val M = st.nextToken().toInt
      val table = mutable.HashMap.empty[(Int, Int), Int]
      val smallestEnd = N / 2 + 1

      def solve(index: Int, startNum: Int): Int = {

        val key = (index, startNum)
        table.get(key) match {
          case Some(ans) => return ans
          case None => {
            var count = 0
            if (index == M) {
              count = N - max(smallestEnd, startNum) + 1
            } else {
              var num = startNum
              while (num * 2 <= N) {
                count = (count + solve(index + 1, num * 2)) % mod
                num += 1
              }
              count = (count + (N - num + 1) * solve(index + 1, 1)) % mod
            }

            table(key) = count
            return count
          }
        }
      }

      println(solve(1, 1))
//      for ((key, value) <- table) {
//        print(key + ": " + value)
//        println()
//      }
    }


  }

  def main(arguments: Array[String]) =
    solve(if (arguments.length == 0) new InputStreamReader(System.in) else new FileReader(arguments(0)))
}
